compiler/modules/front/codeGen.bal

```diff
@@ -312,11 +328,15 @@ function codeGenOnPanic(CodeGenContext cx, Position pos) {
     }
 }
 
-function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialEnv, s:StmtBlock block) returns CodeGenError|StmtEffect {
+// If block of stmts doesn't complete normally, will return empty narrowings and assignments.
+function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialEnv, s:StmtBlock block, StmtNarrowing[] initialNarrowings = []) returns CodeGenError|StmtEffect {
     bir:BasicBlock? curBlock = bb;
     Environment env = environmentCopy(initialEnv);
     final int startRegister = cx.nextRegisterNumber();
-    foreach var stmt in block.stmts {
+    applyEffect(env, { block: bb, narrowings: initialNarrowings });
+    codeGenNarrowings(cx, bb, env, initialNarrowings, block.startPos);
+    int lastStmt = block.stmts.length() - 1;
+    foreach var [i, stmt] in block.stmts.enumerate() {
```

Comment:
Let's call it stmtIndex.

enumerate here is pretty but gratuitously inefficient for something that is used for every statement.